test/test_tree_of_boxes.py

```diff
@@ -194,6 +197,31 @@ def test_tree_plot():
 
 # {{{ test_traversal_from_tob
 
+
+@dataclass
+class TreeOfBoxesForTraversal(TreeOfBoxes):
```

Comment:
Wait, I thought at one point there was a test to ensure that a traversal could be built off of `TreeOfBoxes`. Is that no longer the case?

Reply:
@inducer The code in `test_traversal_for_tob` was setting a lot of additional attributes on `TreeOfBoxes`
https://github.com/inducer/boxtree/blob/1e4123cc08a55478ee715f0757895779c18444aa/test/test_tree_of_boxes.py#L218-L227

I added this class in an attempt to make the whole thing a bit more immutable (#56 makes `Tree` a frozen dataclass and it seems to work). Was it supposed to work in some other way?

Reply:
Oh yikes. I don't know how I missed that. That should not be happening. The traversal build should be able to complete without accessing attributes related to particles. (like `particle_id_dtype`, `sources_are_targets`). The remainder should be available unconditionally on `TreeOfBoxes`. `{sources,targets}_have_extent` is a borderline case. Ideally it should be renamed (so that it's about source/target boxes having extents different from their 'native' ones) and the old name deprecated. But I could see postponing that.

Reply:
All of those things seem to be required to build some kernels and things in the traversal:
https://github.com/inducer/boxtree/blob/1e4123cc08a55478ee715f0757895779c18444aa/boxtree/traversal.py#L1903-L1915

Maybe that's a different PR? Since it seems to be more work and more controversial.

Reply:
Gave this a go in [73cc7de](https://github.com/inducer/boxtree/pull/71/commits/73cc7decf7bafb7e8d95e18d232890ded3f03bd5).

It's probably not complete though (even if the test seems to pass), so help is very welcome. One thing that is missing for sure is that the `box_flags` and `level_start_box_nrs` aren't updated on refinement / coarsening!

Reply:
Thanks! It's better than it was, so I'm good with merging this, but it's overall still an unappealing mess.